drivers/fpga/fpga_eos_s3.c

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2021 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <device.h>
+#include <drivers/fpga.h>
+#include "fpga_eos_s3.h"
+
+void eos_s3_fpga_enable_clk(void)
+{
+	CRU->C16_CLK_GATE = C16_CLK_GATE_PATH_0_ON;
+	CRU->C21_CLK_GATE = C21_CLK_GATE_PATH_0_ON;
+	CRU->C09_CLK_GATE = C09_CLK_GATE_PATH_1_ON | C09_CLK_GATE_PATH_2_ON;
+	CRU->C02_CLK_GATE = C02_CLK_GATE_PATH_1_ON;
+}
+
+void eos_s3_fpga_disable_clk(void)
+{
+	CRU->C16_CLK_GATE = C16_CLK_GATE_PATH_0_OFF;
+	CRU->C21_CLK_GATE = C21_CLK_GATE_PATH_0_OFF;
+	CRU->C09_CLK_GATE = C09_CLK_GATE_PATH_1_OFF | C09_CLK_GATE_PATH_2_OFF;
+	CRU->C02_CLK_GATE = C02_CLK_GATE_PATH_1_OFF;
+}
+
+struct quickfeather_fpga_data {
+	char *FPGA_info;
+};
+
+static enum FPGA_status eos_s3_fpga_get_status(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	if (PMU->FB_STATUS == FPGA_STATUS_ACTIVE) {
+		return FPGA_STATUS_ACTIVE;
+	} else
+		return FPGA_STATUS_INACTIVE;
+}
+
+static const char *eos_s3_fpga_get_info(const struct device *dev)
+{
+	struct quickfeather_fpga_data *data = dev->data;
+
+	return data->FPGA_info;
+}
+
+static int eos_s3_fpga_on(const struct device *dev)
+{
+	if (eos_s3_fpga_get_status(dev) == FPGA_STATUS_ACTIVE) {
+		return 0;
+	}
+
+	PMU->FFE_FB_PF_SW_WU = PMU_FFE_FB_PF_SW_WU_FB_WU;
+	while (PMU->FFE_FB_PF_SW_WU == PMU_FFE_FB_PF_SW_WU_FB_WU) {
+	};
```

Comment:
If we are busy-waiting here, maybe a comment saying why/what for would be nice have here.

Reply:
I added a comment. We are waiting here for the register responsible for waking up the FPGA power domain to be cleared.